tests/regression/src/lib.rs

```diff
@@ -0,0 +1,143 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use s2n_tls::{config::Builder, security, testing::{CertKeyPair, InsecureAcceptAllCertificatesHandler}};
+type Error = s2n_tls::error::Error;
+
+
+
+// Function to create default config with specified parameters
+pub fn set_config(
+    cipher_prefs: &security::Policy,
+    keypair: CertKeyPair
+) -> Result<s2n_tls::config::Config, Error> {
+    let mut builder = Builder::new();
+    builder
+        .set_security_policy(cipher_prefs)
+        .expect("Unable to set config cipher preferences");
+    builder
+        .set_verify_host_callback(InsecureAcceptAllCertificatesHandler {})
+        .expect("Unable to set a host verify callback.");
+    builder
+        .load_pem(keypair.cert(), keypair.key())
+        .expect("Unable to load cert/pem");
+    builder.trust_pem(keypair.cert()).expect("load cert pem");
+    builder.build()
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crabgrind as cg;
+    use s2n_tls::testing::TestPair;
+    use std::env;
+    use std::process::Command;
+
+    const COST: u64 = 1_000_000; //configurable threshold for regression
+    // environment variable to determine whether to run under valgrind or solely test functionality
+    fn is_running_under_valgrind() -> bool {
+        env::var("VALGRIND").is_ok()
+    }
+    //test to create new config, set security policy, host_callback information, load/trust certs, and build config
+    #[test]
+    fn test_set_config() -> Result<(), s2n_tls::error::Error> {
+        if !is_running_under_valgrind() {
+            cg::cachegrind::stop_instrumentation();
+            cg::cachegrind::start_instrumentation();
+            let keypair_rsa = CertKeyPair::default();
+            let _config = set_config(&security::DEFAULT_TLS13, keypair_rsa).expect("Failed to build config");
+            return Ok(());
+        }
+        run_valgrind_test("test_set_config");
+        Ok(())
+    }
+
+    #[test]
+    fn test_rsa_handshake() -> Result<(), s2n_tls::error::Error> {
+        if !is_running_under_valgrind() {
+            cg::cachegrind::stop_instrumentation();
+            // Example usage with RSA keypair (default)
+            let keypair_rsa = CertKeyPair::default();
+            let config = set_config(&security::DEFAULT_TLS13, keypair_rsa)?;
+            // Create a pair (client + server) using that config, start handshake measurement
+            let mut pair = TestPair::from_config(&config);
+            // Assert a successful handshake
+            cg::cachegrind::start_instrumentation();
+            assert!(pair.handshake().is_ok());
+            cg::cachegrind::stop_instrumentation();
+            return Ok(());
+        }
+        run_valgrind_test("test_rsa_handshake");
+        Ok(())
+    }
```

Comment:
Can we abstract this a bit? it seems a bit fragile to me. I'd prefer something like

```rust
#[test]
fn test_rsa_handshake() {
    valgrind_test("test_rsa_handshake", |ctrl| {
        ctl.stop_instrumentation();
        // Example usage with RSA keypair (default)
        let keypair_rsa = CertKeyPair::default();
        let config = set_config(&security::DEFAULT_TLS13, keypair_rsa)?;
        // Create a pair (client + server) using that config, start handshake measurement
        let mut pair = TestPair::from_config(&config);
        // Assert a successful handshake
        ctl.start_instrumentation();
        assert!(pair.handshake().is_ok());
        ctl.stop_instrumentation();
        return Ok(())
    }).unwrap();
}
```

Reply:
Yes, I'll bastract the tests. Why use a closure, though? Is it just for readability, or is there a functional reason for it?

Reply:
Yeah the reason here is it's explicitly defining a contract between the harness and the harness infrastructure. Like, it seems like we leak a lot of _how_ the harness is being executed, rather than just defining _what_ to execute. Concretely, each harness shouldn't have to know to check `is_running_under_valgrind` and then calling `run_valgrind_test(name)`. It should just say: "here's a function I'd like to measure" and the underlying framework takes care of it. Maybe we decide not to use valgrind in the future. Maybe we want to run it multiple times and see if it's deterministic or not.